lib/utmp.c

```diff
@@ -274,8 +274,14 @@ prepare_utmp(const char *name, const char *line, const char *host,
 	if (NULL != ut) {
 		STRNCPY(utent->ut_id, ut->ut_id);
 	} else {
-		/* XXX - assumes /dev/tty?? */
-		STRNCPY(utent->ut_id, line + 3);
+		size_t line_len, ut_line_maxlen;
+		ut_line_maxlen = countof(utent->ut_id) - 1;
+		line_len = strlen (line);
+		if (ut_line_maxlen >= line_len)
+			STRNCPY(utent->ut_id, line);
+		else
+			STRNCPY(utent->ut_id, \
+			        line + (line_len - ut_line_maxlen));
```

Comment:
This seems bogus.  It's truncating from the start, instead of from the end.  Why would we do that?

Reply:
We should instead error out if the string doesn't fit the buffer.

Reply:
> This seems bogus. It's truncating from the start, instead of from the end. Why would we do that?

Because the last four symbols are more unique then first four, when numbers are added at the end.

> We should instead error out if the string doesn't fit the buffer (if it's appropriate to fail, which I don't know), or keep truncating as always.

Definitely not!
A missing utmp entry is a kind of soft error. Nothing that prevents user to login.
`ut_id` does not have to follow `ut_line`. Trimming `ut_line` is just usual way to fill the `ut_id`. Alternatively, just a random strings could be used, but this would not match other utmp file uses.

Check my [comment](https://github.com/shadow-maint/shadow/pull/1293#issuecomment-3077988615) for more details.

Reply:
I still think that simple strlen together with '+' and '-' is not something complex.
What is more important, it aligns behaviour with other modern software.